### PR TITLE
feat: Added UDTs for disks and ability to give a custom name to a disk - `avm/res/compute/virtual-machine`

### LINKED ISSUE #1436: [AVM Module Issue]: compute/virtual-machine and name of OSDisk and DataDisk partly hardcoded in main.bicep
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Bug

### Module Name

avm/res/compute/virtual-machine

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

When I want to use the module in existing pipeline and with existing resources i have the problem with OSDisk and Datadisk name are hardcoded in the bicep file.

image

and they are not to be changed on existing resources.

So you have to have the possibility to configure them with parameters.

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by johnlokerse**
## Description

This pull request adds UDTs for OS disk and data disk. Also, included the option to give a custom name to either disk.

## Pipeline Reference

| Pipeline |
| -------- |
|     [avm.res.compute.virtual-machine](https://github.com/johnlokerse/bicep-registry-modules/actions/workflows/avm.res.compute.virtual-machine.yml)     |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [ ] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [ ] I have run `Set-AVMModule` locally to generate the supporting module files.
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

**Comment by johnlokerse** [comment]
@AlexanderSehr @rahalan 

While running tests I got this error which I can't place. It looks good to me, name is optional and it is checked with coalesce and it still expecting the `name` property. Do you have any idea what this can be? 

image

File: avm/res/compute/virtual-machine/main.bicep [hunk 1]
```diff
@@ -453,6 +453,7 @@ module vm_nic 'modules/nic-configuration.bicep' = [
 resource vm 'Microsoft.Compute/virtualMachines@2022-11-01' = {
   name: name
   location: location
+  #disable-next-line BCP036
```
Comment (johnlokerse) [inline avm/res/compute/virtual-machine/main.bicep:456]: Trying to fix this later. (this is already in main @rahalan )

Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep:456]: @johnlokerse, not sure I understand the comment? Is there something I can help?

File: avm/res/compute/virtual-machine/main.bicep [hunk 2]
```diff
@@ -474,35 +475,27 @@ resource vm 'Microsoft.Compute/virtualMachines@2022-11-01' = {
     storageProfile: {
       imageReference: imageReference
       osDisk: {
-        name: '${name}-disk-os-01'
-        createOption: contains(osDisk, 'createOption') ? osDisk.createOption : 'FromImage'
-        deleteOption: contains(osDisk, 'deleteOption') ? osDisk.deleteOption : 'Delete'
+        name: osDisk.name ?? '${name}-disk-os-01'
```
Comment (rahalan) [inline avm/res/compute/virtual-machine/main.bicep:478]: I think you need to check first, if name property is set (with contains, see old code)

Comment (johnlokerse) [reply avm/res/compute/virtual-machine/main.bicep:478]: Hey @rahalan - thanks for thinking with me 😃. Got it working, forgot to add the safe-dereference operator (`?`): https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/operators#safe-dereference. Some of the user-defined type properties are nullable so they need to be accessed with the safe-dereference operator else the build/deployment breaks.

I used coalesce (`??`) operator because I find it cleaner than the `contains()` function. Also, it is used this way in other Bicep AVM templates. I am OK with using contains() if you find that more suitable for this template?

File: avm/res/compute/virtual-machine/main.bicep (hunk 1 above)
Comment (johnlokerse) [reply avm/res/compute/virtual-machine/main.bicep:456]: Sorry for the confusion. I came across this error which was already in the main branch and it was a reminder for me to try and fix it. Just tagged you for visibility :-)

Comment (AlexanderSehr) [reply avm/res/compute/virtual-machine/main.bicep]: When disabling I'd in general recommend to add a comment. These identifiers are rather cryptic after all 😄

File: avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep [hunk 3]
```diff
@@ -12,7 +12,8 @@ metadata description = 'This instance deploys the module for a VM with dedicated
 param resourceGroupName string = 'dep-${namePrefix}-compute.virtualMachines-${serviceShort}-rg'
 
 @description('Optional. The location to deploy resources to.')
-param resourceLocation string = deployment().location
+#disable-next-line no-unused-params // required because Test-TemplateDeployment.ps1 requires the resourceLocation parameter, value must equal to tempLocation
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep:15]: Instead I'd actually suggtest to just assign the value to the resourceGroup. Doesn't change the functionality and saves us an extra suppresion :)

File: avm/res/compute/virtual-machine/main.bicep [hunk 4]
```diff
@@ -1060,3 +1053,80 @@ type roleAssignmentType = {
...
+      | 'PremiumV2_LRS'
+      | 'Premium_LRS'
+      | 'Premium_ZRS'
+      | 'StandardSSD_LRS'
+      | 'StandardSSD_ZRS'
+      | 'Standard_LRS'
+      | 'UltraSSD_LRS'
+
+    @description('Optional. Specifies the customer managed disk encryption set resource id for the managed disk.')
+    diskEncryptionSet: {
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/main.bicep:1127]: Usually, for cases where only a single ID entry exists, we'd opt to change the input object to a `diskEncryptionSetResourceId` and create the object for the user. While it divers slightly from the RP, it's easier to use, and more descriptive (be it only because it explicitly calls for a resourceId and nothering else). Same for the osDisk. Thoughts?

File: avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep (hunk 3 above)
Comment (johnlokerse) [reply avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep]: Done!

File: avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep [hunk 5]
```diff
@@ -11,8 +11,8 @@ metadata description = 'This instance deploys the module for a VM with dedicated
 @maxLength(90)
 param resourceGroupName string = 'dep-${namePrefix}-compute.virtualMachines-${serviceShort}-rg'
 
-@description('Optional. The location to deploy resources to.')
-param resourceLocation string = deployment().location
+@description('Optional. The location to deploy resources to.') // required because Test-TemplateDeployment.ps1 requires the resourceLocation parameter, value must equal to tempLocation
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep:14]: Similar suggestion as per the windows hostpool test

File: avm/res/compute/virtual-machine/main.bicep (hunk 4 above)
Comment (johnlokerse) [reply avm/res/compute/virtual-machine/main.bicep:1127]: I agree. It saves the user some object complexity and makes it more straight-forward. Replaced this in the osDisk and dataDisk.

File: avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep (hunk 3 above)
Comment (johnlokerse) [reply avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep]: Done!

File: avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep (hunk 5 above)
Comment (johnlokerse) [reply avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep:14]: Done!

File: avm/res/compute/virtual-machine/main.bicep (hunk 1 above)
Comment (johnlokerse) [reply avm/res/compute/virtual-machine/main.bicep]: Fixed it, was a small typo in the variable. A "space" was missing:

image

**Comment by johnlokerse** [comment]
Implemented PR feedback. Running Actions for tests now. 

@rahalan Yesterday I had a capacity error for the Nvidia test. What region do you suggest to set when I get this error again?

**Comment by rahalan** [comment]
> Implemented PR feedback. Running Actions for tests now.
> 
> @rahalan Yesterday I had a capacity error for the Nvidia test. What region do you suggest to set when I get this error again?

You can't. Just ignore the test. To get it running you need to ask for a quota of the VM type.

**Comment by johnlokerse** [comment]
> > Implemented PR feedback. Running Actions for tests now.
> > @rahalan Yesterday I had a capacity error for the Nvidia test. What region do you suggest to set when I get this error again?
> 
> You can't. Just ignore the test. To get it running you need to ask for a quota of the VM type.

Ok will do!

@AlexanderSehr @rahalan How do we deal with key vaults that already exist somewhere in the Azure world? Deployment name could be used with a unique string.

**Comment by rahalan** [comment]
@johnlokerse because of that e have the TOKEN_NAMEPREFIX. Maybe try a different one.

**Comment by rahalan** [comment]
@johnlokerse I can't see the actions in your fork, thus cannot see the workflow run. Can you please check the permissions on that? Thanks!

**Comment by AlexanderSehr** [comment]
> @johnlokerse I can't see the actions in your fork, thus cannot see the workflow run. Can you please check the permissions on that? Thanks!

My best guess is that the actions are simply disabled.

**Comment by rahalan** [comment]
> > @johnlokerse I can't see the actions in your fork, thus cannot see the workflow run. Can you please check the permissions on that? Thanks!
> 
> My best guess is that the actions are simply disabled.

I don't think so, as a link to the run is at the top of this PR. I only get a 404 when clicking it.

**Comment by johnlokerse** [comment]
@AlexanderSehr @rahalan I have made several contributions to AVM and once made a mistake which caused some workflows to deploy to my personal Azure environment without me knowing and generating costs. This is why I disable the workflows when I am not working on them. 

I have enabled the actions now. I am seeing unreliable test deployments, for example, sometimes I get 404 errors, key vault name already in use, resources that cannot be deleted in the post deployment stage and so on. Every test deployment is different so I can't reproduce it. Still got some work to do before merging 😄

**Comment by AlexanderSehr** [comment]
> @AlexanderSehr @rahalan I have made several contributions to AVM and once made a mistake which caused some workflows to deploy to my personal Azure environment without me knowing and generating costs. This is why I disable the workflows when I am not working on them.
> 
> I have enabled the actions now. I am seeing unreliable test deployments, for example, sometimes I get 404 errors, key vault name already in use, resources that cannot be deleted in the post deployment stage and so on. Every test deployment is different so I can't reproduce it. Still got some work to do before merging 😄

Hey @johnlokerse,
this sadly makes 110 % sense - but - there is good news
- Regarding the 'xzy already in use' error I recommend to change the namePrefix secret to something even more ludicrous
- Regarding the costs, there is a workflow mentioned in the tip box [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/#33-set-readwrite-workflow-permissions) - but in short: I highly recommend to run the `avm.platform.toggle-avm-workflows.yml` workflow, which will essentially allow you to disable/enable workflows in one swift swoop (by default disable all res & ptn workflows)

Otherwise the schedule that runs once a week will indeed eat your credits away like its candy.

**Comment by johnlokerse** [comment]
> > @AlexanderSehr @rahalan I have made several contributions to AVM and once made a mistake which caused some workflows to deploy to my personal Azure environment without me knowing and generating costs. This is why I disable the workflows when I am not working on them.
> > I have enabled the actions now. I am seeing unreliable test deployments, for example, sometimes I get 404 errors, key vault name already in use, resources that cannot be deleted in the post deployment stage and so on. Every test deployment is different so I can't reproduce it. Still got some work to do before merging 😄
> 
> Hey @johnlokerse, this sadly makes 110 % sense - but - there is good news
> 
> * Regarding the 'xzy already in use' error I recommend to change the namePrefix secret to something even more ludicrous
> * Regarding the costs, there is a workflow mentioned in the tip box [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/#33-set-readwrite-workflow-permissions) - but in short: I highly recommend to run the `avm.platform.toggle-avm-workflows.yml` workflow, which will essentially allow you to disable/enable workflows in one swift swoop (by default disable all res & ptn workflows)
> 
> Otherwise the schedule that runs once a week will indeed eat your credits away like its candy.

Didn't know about the toggle to enable/disable workflows. This is great! Will take a look at the namePrefix, thanks 😄 

@rahalan @AlexanderSehr Did you also encounter this error? https://github.com/johnlokerse/bicep-registry-modules/actions/runs/8590236271/job/23537534815#step:4:1187
I think we probably are missing a check somewhere in Invoke-ResourcePostRemoval.ps1.

**Comment by AlexanderSehr** [comment]
I can't say I have, but you're right it looks like it. 🤔

**Comment by rahalan** [comment]
@johnlokerse I can see your workflow runs now, thanks.

**Comment by johnlokerse** [comment]
Green workflow (except for Nvidia):
https://github.com/johnlokerse/bicep-registry-modules/actions/runs/8687076991

It didn't include the static validation step because I replaced the value of `#_namePrefix_#` for testing purposes... builds without the `#_namePrefix_#` are rejected. So I tried to rerun it with the original value (`#_namePrefix_#`) and it failed at the deployment of the key vault due to the name that is not globally unique:
https://github.com/johnlokerse/bicep-registry-modules/actions/runs/8687839244

Would you like me to wait until I can run all the steps for a green workflow?

**Comment by rahalan** [comment]
@johnlokerse regarding the nameprefix: the name prefix needs to be a secret in your GH, see 3.1 https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/#31-set-up-secrets

**Review by rahalan — APPROVED** [review]

**Comment by johnlokerse** [comment]
> @johnlokerse regarding the nameprefix: the name prefix needs to be a secret in your GH, see 3.1 https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/#31-set-up-secrets

@rahalan Oh right completely forgot about that... will run the tests with updated prefix tomorrow. FYI; Also need to revert the backup service guid

File: avm/res/compute/virtual-machine/tests/e2e/windows.hostpool/main.test.bicep [hunk 6]
```diff
@@ -24,6 +24,10 @@ param password string = newGuid()
 @description('Optional. A token to inject into the name of each resource.')
 param namePrefix string = '#_namePrefix_#'
 
+// Set to fixed location as the RP function returns unsupported locations
+// Right now (2024/04) the following locations are supported: centralindia,uksouth,ukwest,japaneast,australiaeast,canadaeast,canadacentral,northeurope,westeurope,eastus,eastus2,westus,westus2,westus3,northcentralus,southcentralus,westcentralus,centralus
+param enforcedLocation string = 'westeurope'
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/windows.hostpool/main.test.bicep:29]: @rahalan we should **not** use West Europe. Please change in your next PR

File: avm/res/compute/virtual-machine/tests/e2e/atmg/main.test.bicep [hunk 7]
```diff
@@ -20,6 +20,10 @@ param serviceShort string = 'cvmlinatmg'
 @description('Optional. A token to inject into the name of each resource.')
 param namePrefix string = '#_namePrefix_#'
 
+// Set to fixed location as the RP function returns unsupported locations (configurationProfileAssignments)
+// Right now (2024/04) the following locations are supported: centralus, eastus, eastus2, southcentralus, westus, westus2, westcentralus, northeurope, westeurope, canadacentral, japaneast, uksouth, australiasoutheast, australiaeast, southeastasia, westus3
+param enforcedLocation string = 'westeurope'
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/atmg/main.test.bicep:25]: @rahalan we should not use West Europe. Please change in your next PR

File: avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep [hunk 8]
```diff
@@ -35,7 +35,7 @@ var tempLocation = 'eastus'
 // =================
 resource resourceGroup 'Microsoft.Resources/resourceGroups@2021-04-01' = {
   name: resourceGroupName
-  location: tempLocation
+  location: resourceLocation
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/windows.nvidia/main.test.bicep:38]: Why was this changed from the hardcoded location? @rahalan I suggest to undo this in your next PR

File: avm/res/compute/virtual-machine/tests/e2e/windows.max/dependencies.bicep [hunk 9]
```diff
@@ -215,10 +217,10 @@ resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' = {
 }
 
 resource backupServiceKeyVaultPermissions 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
-  name: guid('${keyVault.name}-${location}-268f6a53-9f68-4a38-ae47-166f730d86af-KeyVault-KeyVaultAdministrator-RoleAssignment')
+  name: guid('${keyVault.name}-${location}-94d1f9d9-9caf-4e06-809e-29fc95f69e65-KeyVault-KeyVaultAdministrator-RoleAssignment')
   scope: keyVault
   properties: {
-    principalId: '268f6a53-9f68-4a38-ae47-166f730d86af' // Backup Management Service Enterprise Application Object Id (Note: this is tenant specific)
+    principalId: '94d1f9d9-9caf-4e06-809e-29fc95f69e65' // Backup Management Service Enterprise Application Object Id (Note: this is tenant specific)
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/windows.max/dependencies.bicep:223]: @rahalan as the comment says, this value is tenant specific and should have been changed back before merging the PR. The workflow also broke accordingly. Please change in your next PR.

File: avm/res/compute/virtual-machine/tests/e2e/waf-aligned/dependencies.bicep [hunk 10]
```diff
@@ -215,10 +217,10 @@ resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' = {
 }
 
 resource backupServiceKeyVaultPermissions 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
-  name: guid('${keyVault.name}-${location}-268f6a53-9f68-4a38-ae47-166f730d86af-KeyVault-KeyVaultAdministrator-RoleAssignment')
+  name: guid('${keyVault.name}-${location}-94d1f9d9-9caf-4e06-809e-29fc95f69e65-KeyVault-KeyVaultAdministrator-RoleAssignment')
   scope: keyVault
   properties: {
-    principalId: '268f6a53-9f68-4a38-ae47-166f730d86af' // Backup Management Service Enterprise Application Object Id (Note: this is tenant specific)
+    principalId: '94d1f9d9-9caf-4e06-809e-29fc95f69e65' // Backup Management Service Enterprise Application Object Id (Note: this is tenant specific)
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/waf-aligned/dependencies.bicep:223]: @rahalan as the comment says, this value is tenant specific and should have been changed back before merging the PR. The workflow also broke accordingly. Please change in your next PR.

File: avm/res/compute/virtual-machine/tests/e2e/linux.max/dependencies.bicep [hunk 11]
```diff
@@ -218,10 +220,10 @@ resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' = {
 }
 
 resource backupServiceKeyVaultPermissions 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
-  name: guid('${keyVault.name}-${location}-268f6a53-9f68-4a38-ae47-166f730d86af-KeyVault-KeyVaultAdministrator-RoleAssignment')
+  name: guid('${keyVault.name}-${location}-94d1f9d9-9caf-4e06-809e-29fc95f69e65-KeyVault-KeyVaultAdministrator-RoleAssignment')
   scope: keyVault
   properties: {
-    principalId: '268f6a53-9f68-4a38-ae47-166f730d86af' // Backup Management Service Enterprise Application Object Id (Note: this is tenant specific)
+    principalId: '94d1f9d9-9caf-4e06-809e-29fc95f69e65' // Backup Management Service Enterprise Application Object Id (Note: this is tenant specific)
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/linux.max/dependencies.bicep:226]: @rahalan as the comment says, this value is tenant specific and should have been changed back before merging the PR. The workflow also broke accordingly. Please change in your next PR.

**Comment by AlexanderSehr** [comment]
Hey @johnlokerse, I could help but also take a peak as I was mentioned in a comment (as per my notifications). Commented a couple other things aside the GUID that you already mentioned that I'd suggest to change/fix soon.

**Comment by johnlokerse** [comment]
@AlexanderSehr @rahalan Yes I will pick this up, will be my first thing to do tomorrow! 

Didn't expect a merge this quick, so my bad.

File: avm/res/compute/virtual-machine/tests/e2e/atmg/main.test.bicep (hunk 7 above)
Comment (johnlokerse) [reply avm/res/compute/virtual-machine/tests/e2e/atmg/main.test.bicep:25]: @AlexanderSehr Are there any guidelines on what region must be used in this situation?

Comment (johnlokerse) [reply avm/res/compute/virtual-machine/tests/e2e/atmg/main.test.bicep:25]: In other Bicep templates the `enforcedLocation` variable is scoped to either `eastus` or `eastasia`.  So eastus would be the choice here. Is that correct?

**Comment by rahalan** [comment]
@johnlokerse @AlexanderSehr I didn't review good enough, my bad. I will change the GUID back to the correct one in my already open PR
